Implement HSTS fetch step

nmvk opened the pull request:
Implemented step nine of the main fetch. If current URL scheme is 'HTTP' and
current URL's host is domain and if current URL's host matched with Known
HSTS Host Domain Name Matching results in either a superdomain match with
an asserted includeSubDomains directive or a congruent match then we
change request scheme to 'https'. This change has been made in method.rs

A test case to validate this has been added in fetch.rs. For asserting
https scheme, a https localhost was required. For this purpose I have
created a self-signed certificate and refactored fetch-context and
connector.rs to programmatically trust this certificate for running this
test case.

This should fix https://github.com/servo/servo/issues/14363
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14363 

<!-- Either: -->
- [X] There are tests for these changes

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14716)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/fetch/methods.rs, components/net/resource_thread.rs, components/net/http_loader.rs, components/net/connector.rs

mrnayak commented:
Travis failure due to `The command "./mach test-stylo" exited with 101.` not due to the current change.

jdm commented:
That's #14723 .

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14741) made this pull request unmergeable. Please resolve the merge conflicts.

mrnayak commented:
I have addressed conflicts. @emilio did you get a chance to review the changes?

emilio commented:
I think @jdm or @Ms2ger are more suited to review this than me.

There should probably be a WPT test or similar for this.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit e8099df with merge 1ccb92f...

emilio commented:
@bors-servo r? @Ms2ger

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1607)

mrnayak commented:
Is the test failure due to network issue?

emilio commented on components/net/fetch/methods.rs:
```diff
@@ -189,7 +189,15 @@ pub fn main_fetch(request: Rc<Request>,
 
     // Step 9
-    // TODO this step (HSTS)
+    if request.current_url().scheme() == "http" && request.current_url().domain() != None {
```
nit: use `request.current_url().domain().is_some()`?

emilio reviewed (commented):
Seems like the failures may be legit? We can retry to verify it.

Other option would be that that check makes stuff immensely slower, in which case we'd need a faster data-structure for HstsList.

emilio commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit e8099df with merge 35d62a5...

emilio commented:
IMO the hsts list seems pretty big to use a vec as a backing store, but maybe other people have different opinions.

mrnayak commented:
Thanks, @emilio, Maybe I could refactor `HstsList` to use `HashSet` or `HashMap` which could provide lookup?

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4637), [arm32](http://build.servo.org/builders/arm32/builds/4812), [arm64](http://build.servo.org/builders/arm64/builds/4836), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4926), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1602), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1610), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4903), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4943), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1610), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1644), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/65), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/66)

emilio commented:
Tests seem green, so it's likely to have been another kind of error. I've opened #14756 to change the underlying storage of the HSTS list. I think it's worth to land that refactoring separately as a followup.

jdm commented on resources/certificate.crt:
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
```
This file should be called `self_signed_certificate_for_testing.crt` to make its purpose clear ;)

jdm commented on resources/privatekey.key:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
This should be called `privatekey_for_testing.key`.

jdm commented on tests/unit/net/fetch.rs:
```diff
@@ -506,6 +512,55 @@ fn test_fetch_with_local_urls_only() {
+        list.push(HstsEntry::new("localhost".to_owned(), IncludeSubdomains::NotIncluded, None)
+            .unwrap());
+    }
+    let do_fetch = |url: ServoUrl| {
```
I don't think having this closure makes the test any more readable, so let's remove it.

jdm commented on tests/unit/net/fetch.rs:
```diff
@@ -506,6 +512,55 @@ fn test_fetch_with_local_urls_only() {
+fn test_fetch_with_hsts() {
+    static MESSAGE: &'static [u8] = b"";
+    let handler = move |_: HyperRequest, response: HyperResponse| {
+        response.send(MESSAGE).unwrap();
```
Let's have the response contain the scheme of the request's URL instead. This will let us verify that the server never receives a non-HTTPS connection.

mrnayak replied:
If I correctly understand, you want `request.url.scheme()` to be sent as part of response body?. Here we use `hyper::server::request::Request` which does not capture information about request scheme. If we make a non-HTTPS connection to the server, then we get response error saying "Connection reset"

jdm replied:
Ah, good point. This test is fine as-is, in that case.

jdm reviewed (changes requested):
Great work!

nmvk committed:
Implement HSTS fetch step

Implemented step nine of the main fetch. If current URL scheme is 'HTTP'
and current URL's host is domain and if current URL's host matched with
Known
HSTS Host Domain Name Matching results in either a superdomain match with
an asserted includeSubDomains directive or a congruent match then we
change request scheme to 'https'. This change has been made in method.rs

A test case to validate this has been added in fetch.rs. For asserting
https scheme, a https localhost was required. For this purpose I have
created a self-signed certificate and refactored fetch-context and
connector.rs to programmatically trust this certificate for running this
test case.

jdm commented:
@bors-servo: r+
Thanks for fixing this and figuring out how to use HTTPS servers in unit tests!

bors-servo commented:
:pushpin: Commit 6020b4c has been approved by `jdm`

<!-- @bors-servo r=jdm 6020b4c15c62f950346d6e446fb9a8ed04d05e13 -->

bors-servo commented:
:hourglass: Testing commit 6020b4c with merge c7991d5...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4652), [arm32](http://build.servo.org/builders/arm32/builds/4826), [arm64](http://build.servo.org/builders/arm64/builds/4851), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4941), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1617), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1625), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4918), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4958), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1625), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1659), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/80), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/81)
